### DISCUSSION TITLE
different boot menus for different IP addresses or network ranges

### DISCUSSION THREAD

**Discussion by CHANLE-01** [ipxe/ipxe#1353] [Q&A]
Hi all,
I have iPXE systems with WDS (Windows) or standard DHCP/TFTPD (Linux) up and running.
the iPXEfiles (snponly_64bit.efi etc.) are precompiled by the vendor of the software we use and starts with ipxe.menu (the boot menu).
Everything works fine, we are doing a few thousands of network boots a day on different locations.
-
But now i need to add more independent network ranges and boot menus  to the existing systems.
How to do this ?
compile the iPXE files myself, set another start file for each subnet, let DHCP do the assignment.
Or
is it possible to have in the existing ipxe.menu some settings to chain load a sub menu ?
this will be my preferred solution, but i did not find any hint how to do this.

With legacy PXE boot it was easy, i had to rename the pxelinux.cfg file to eg. A0A000 and then only 10.10.0.x clients got the menu.

Maybe someone can help me and show me the right direction where to go.

Regards
CH

**Comment by NiKiZe** [comment]
As long as you have access to modify any of the iPXE scripts that is used you can do what you want.

Without knowing a lot more about the current system it is hard to give suggestions, even then it is probably out of scope for this forum.

With that said, try in a separate environment and create some kind of menu or boot system that at least starts to do what you want, and then try to move that in to the existing setup once you have a better understanding of where and how you want to use your modifications.

**Reply by CHANLE-01** [reply]
maybe this explains my requirement much better:
I have eg 3 vlans with this 3 address ranges:
10.249.10.0/22, 10.249.20.0/22, 10.249.30.0/22.
My existing implementation executes my ipxe.menu, all clients get the same menu.
All good, it works, the users can manually choose submenus

But now i must have for each address range an independent menu.
if client is in 10.249.10.0/22 range he shall load ipxe10.menu
if client is in 10.249.20.0/22 range he shall load ipxe20.menu
if client is in 10.249.30.0/22 range he shall load ipxe30.menu
etc.

Maybe it can be done like this:
show ip gives me as example: nrt0.dhcp/ip:ipv4 = 10.249.10.11.
an then:
if IP address contains 10.249.1 then chainload pxe10.menu
if IP address contains 10.249.2 then chainload pxe20.menu
if IP address contains 10.249.3 then chainload pxe30.menu
no interaction from the user required to load the proper (sub)menu

I need a hint or example how to implement this.

Regards
CH

**Reply by NiKiZe** [reply]
Either set your dhcp to return the filename relevant for the vlan in question.

Another more dynamic approach is to load the scripts over http and let the server return the content relevant for the given source address of the request.

**Comment by skyblaster** [comment]
Is there no way to parse the individual octets from the current DHCP issued IP address ${net0/ip}?

Something like the following would be great if possible:
```
iseq ${second-octet} 101 && goto server1 ||
iseq ${second-octet} 102 && goto server2 ||
goto default
```

**Reply by NiKiZe** [reply]
Intentionally not supported, see https://github.com/ipxe/ipxe/discussions/326#discussioncomment-629914

Just send this of to a server, use DNS, and let the server give you the correct script.

**Reply by skyblaster** [reply]
Thanks!

EDIT: I'll leave this CGI example here for reference as it works quite well with apache2, however the cgi modules are now deprecated in Python 3.13.
```python
#!/usr/bin/env python3
import cgi
import cgitb
import re

cgitb.enable()

print("Content-Type: text/plain")
print()

form = cgi.FieldStorage()
client_ip = form.getvalue("address", "")

ip_pattern = re.compile(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$")
if not ip_pattern.match(client_ip):
    print("#!ipxe")
    print("echo Invalid or missing IP address")
    print("exit")
    exit()

ip_parts = client_ip.split(".")
prefix = f"{ip_parts[0]}.{ip_parts[1]}"

servers = {
    "10.101": "server1.example.com",
    "10.102": "server2.example.com"
}

server = servers.get(prefix)

print("#!ipxe")
if server:
    print("dhcp net0")
    print(f"set boot-url http://{server}")
    print("kernel ${boot-url}/wimboot gui quiet")
    print("initrd ${boot-url}/boot/BCD             BCD")
    print("initrd ${boot-url}/boot/boot.sdi        BOOT.SDI")
    print("initrd ${boot-url}/sources/x64/boot.wim BOOT.WIM")
    print("boot")
else:
    print(f"echo No server configured for IP prefix {prefix}")
    print("exit")
```

**Comment by skyblaster** [comment]
Here is a more modern Flask example. I left the logging section commented as it only worked in development mode. With Gunicorn as the WSGI server, logs did not follow, which I'm okay with.

I intend to run Caddy as a reverse proxy to Gunicorn, in which case the HTTP status codes in the script may be completely unnecessary.

```python
#!/usr/bin/env python3
from flask import Flask, request
import re
import logging
import traceback

app = Flask(__name__)

# Set up logging
#logging.basicConfig(
#    level=logging.ERROR,
#    format='%(asctime)s - %(levelname)s - %(message)s',
#    filename='/var/log/flask_app.log',
#    filemode='a'
#)

# Ensure all responses are text/plain
@app.after_request
def set_content_type(response):
    response.headers['Content-Type'] = 'text/plain'
    return response

def generate_ipxe_error(message):
    return f"""#!ipxe
echo {message}
exit
"""

@app.route('/boot', methods=['GET'])
def generate_ipxe():
    try:
        client_ip = request.args.get('address', '')

        ip_address = re.compile(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$")
        if not ip_address.match(client_ip):
            return generate_ipxe_error("Invalid or missing IP address"), 400

        ip_parts = client_ip.split('.')
        if not all(0 <= int(part) <= 255 for part in ip_parts):
            return generate_ipxe_error("Invalid IP address octets"), 400

        prefix = f"{ip_parts[0]}.{ip_parts[1]}"

        servers = {
            "10.101": "server1.example.com",
            "10.102": "server2.example.com"
        }

        server = servers.get(prefix)
        if not server:
            return generate_ipxe_error(f"No server configured for IP prefix {prefix}"), 404

        return f"""#!ipxe
dhcp net0
set boot-url http://{server}
kernel ${{boot-url}}/wimboot gui quiet
initrd ${{boot-url}}/boot/BCD             BCD
initrd ${{boot-url}}/boot/boot.sdi        BOOT.SDI
initrd ${{boot-url}}/sources/x64/boot.wim BOOT.WIM
boot
""", 200

    except Exception as e:
        logging.error("Flask app error:\n%s", traceback.format_exc())
        return generate_ipxe_error("Internal server error"), 500

if __name__ == '__main__':
    app.run(debug=False)
```
